wt-status-serialize.c

```diff
@@ -4,6 +4,92 @@
 
 static struct trace_key trace_serialize = TRACE_KEY_INIT(SERIALIZE);
 
+/*
+ * Compute header record for exclude file using format:
+ *      <key> SP <status_char> SP <variant> LF
+ */
+void wt_serialize_compute_exclude_header(struct strbuf *sb,
+					 const char *key,
+					 const char *path)
+{
+	struct stat st;
+	struct stat_data sd;
+
+	memset(&sd, 0, sizeof(sd));
+
+	strbuf_setlen(sb, 0);
+
+	if (!path || !*path) {
+		strbuf_addf(sb, "%s U (unset)", key);
+	}
+	else if (lstat(path, &st) == -1) {
+		if (is_missing_file_error(errno))
+			strbuf_addf(sb, "%s E (not-found) %s", key, path);
+		else
+			strbuf_addf(sb, "%s E (other) %s", key, path);
+	}
+	else {
+		fill_stat_data(&sd, &st);
+		strbuf_addf(sb, "%s F %d %d %s",
+			    key, sd.sd_mtime.sec, sd.sd_mtime.nsec, path);
+	}
+}
+
+static void append_exclude_info(int fd, const char *path, const char *key)
+{
+	struct strbuf sb = STRBUF_INIT;
+
+	wt_serialize_compute_exclude_header(&sb, key, path);
+
+	packet_write_fmt(fd, "%s\n", sb.buf);
+
+	strbuf_release(&sb);
+}
+
+static void append_core_excludes_file_info(int fd)
+{
+	/*
+	 * Write pathname and mtime of the core/global excludes file to
+	 * the header since a change in the excludes will change the
+	 * results reported by status.
+	 *
+	 * The "core.excludefile" setting defaults to $XDG_HOME/git/ignore
+	 * and uses a global variable which should have been set during
+	 * wt_status_collect_untracked().
+	 *
+	 * See dir.c:setup_standard_excludes()
+	 */
+	append_exclude_info(fd, excludes_file, "core_excludes");
+}
+
+static void append_repo_excludes_file_info(int fd)
+{
+	/*
+	 * Likewise, there is a per-repo excludes file in .git/info/excludes
+	 * that will change the results reported by status.
+	 *
+	 * See dir.c:setup_standard_excludes() and git_path_info_excludes().
+	 * We replicate the pathname construction here because of the static
+	 * variables/functions used in dir.c.
+	 */
+	char *path = git_pathdup("info/exclude");
+
+	append_exclude_info(fd, path, "repo_excludes");
+
+	free(path);
+}
+
+/*
+ * WARNING: there are also per-directory .gitignore files that can change
+ * the results reported by status.  If you run a serialize with untracked
+ * .gitignore files and then change a per-directory .gitignore file or
```

Comment:
I was confused when I initially read this, but I think I understand it now. `untracked .gitignore files` - I think this is referring to untracked files that match an exclude pattern, rather than `.gitignore` files that are untracked.

Reply:
I probably made the comment more confusing that it needs to be.  

Every directory is allowed to have a .gitignore file.  Their presence can alter the output of status.
If a new .gitignore file is created (after a status snapshot is created) -or- if an existing .gitignore
file is modified (after the snapshot is created), the results of status may change.

The first type I called untracked.  The second type I called unstaged.  The main point is that the
set of what should be excluded/ignored has changed and the index was not updated.  If either
type of .gitignore files are added/staged, the index will be updated and its mtime changed, so
the existing checks will consider the status cache invalid.  My point was that those 2 types of
.gitignore files can exist and the index mtime won't have changed.